memtable/write_buffer_manager.cc

```diff
@@ -0,0 +1,115 @@
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#include "rocksdb/write_buffer_manager.h"
+#include <mutex>
+#include "util/coding.h"
+
+namespace rocksdb {
+#ifndef ROCKSDB_LITE
+namespace {
+const size_t kSizeDummyEntry = 1024 * 1024;
+// The key will be longer than blocks in SST files so it won't conflict.
+const size_t kCacheKeyPrefix = kMaxVarint64Length * 4 + 1;
+}  // namespace
+
+struct WriteBufferManager::CacheRep {
+  std::shared_ptr<Cache> cache_;
+  std::mutex cache_mutex_;
+  std::atomic<size_t> cache_allocated_size_;
+  char cache_key_prefix_[kCacheKeyPrefix];
+  uint64_t num_cache_entries_ = 0;
+  std::vector<Cache::Handle*> dummy_handles_;
+
+  Slice GetCacheKey(char* cache_key, uint64_t id) {
+    memcpy(cache_key, cache_key_prefix_, kCacheKeyPrefix);
+    char* end = EncodeVarint64(cache_key + kCacheKeyPrefix, id);
+    return Slice(cache_key, static_cast<size_t>(end - cache_key));
+  }
+};
+#else
+struct WriteBufferManager::CacheRep {};
+#endif  // ROCKSDB_LITE
+
+WriteBufferManager::WriteBufferManager(size_t _buffer_size,
+                                       std::shared_ptr<Cache> cache)
+    : buffer_size_(_buffer_size),
+      memory_used_(0),
+      mutable_memtable_used_(0),
+      cache_rep_(nullptr) {
+#ifndef ROCKSDB_LITE
+  if (cache) {
+    // Construct the cache key using the pointer to this.
+    cache_rep_.reset(new CacheRep());
+    cache_rep_->cache_ = cache;
+    memset(cache_rep_->cache_key_prefix_, 0, kCacheKeyPrefix);
+    size_t pointer_size = sizeof(const void*);
+    assert(pointer_size <= kCacheKeyPrefix);
+    memcpy(cache_rep_->cache_key_prefix_, static_cast<const void*>(this),
+           pointer_size);
+  }
+#endif  // ROCKSDB_LITE
+}
+
+WriteBufferManager::~WriteBufferManager() {
+#ifndef ROCKSDB_LITE
+  if (cache_rep_) {
+    for (auto* handle : cache_rep_->dummy_handles_) {
+      cache_rep_->cache_->Release(handle, true);
+    }
+  }
+#endif  // ROCKSDB_LITE
+}
+
+// Should only be called from write thread
+void WriteBufferManager::ReserveMemWithCache(size_t mem) {
+#ifndef ROCKSDB_LITE
+  assert(cache_rep_ != nullptr);
+  // Use a mutex to protect various data structures. Can be optimzied to a
+  // lock-free solution if it ends up with a performance bottleneck.
+  std::lock_guard<std::mutex> lock(cache_rep_->cache_mutex_);
+
+  size_t new_mem_used = memory_used_.load(std::memory_order_relaxed) + mem;
+  memory_used_.store(new_mem_used, std::memory_order_relaxed);
+  while (new_mem_used > cache_rep_->cache_allocated_size_) {
+    // Expand size by at least 1MB.
+    // Add a dummy record to the cache
+    char cache_key[kCacheKeyPrefix + kMaxVarint64Length];
+    Cache::Handle* handle;
+    cache_rep_->cache_->Insert(
+        cache_rep_->GetCacheKey(cache_key, cache_rep_->num_cache_entries_++),
+        nullptr, kSizeDummyEntry, nullptr, &handle);
+    cache_rep_->dummy_handles_.push_back(handle);
```

Comment:
The mutex mainly protects the synchronization between the synchronization between memory_used_ and cache_rep_->cache_allocated_size_. If two threads calling ReserveMem() and FreeMem() together, it is not easy to make sure we end up with reserving enough for the memory reserved. I thought about doing it in lockless way for a while and it didn't end up with easy to me.